[WPE] WPEPlatform: add default accessibility implementation using ATK

carlosgcampos opened the pull request:
#### f62eb108139f6b7f4205ae6e6a7b72d7f1ab37a8
<pre>
[WPE] WPEPlatform: add default accessibility implementation using ATK
<a href="https://bugs.webkit.org/show_bug.cgi?id=289329">https://bugs.webkit.org/show_bug.cgi?id=289329</a>

Reviewed by Adrian Perez de Castro.

When WebKit is built with ATK and the platform doesn&apos;t implement
WPEView::get_accessible, a default implementation is provided using ATK.

* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::accessible):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::~View):
(WKWPE::View::accessible const): Deleted.
* Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebViewLegacy.cpp:
(WKWPE::ViewLegacy::~ViewLegacy):
(WKWPE::ViewLegacy::accessible const):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewLegacy.h:
* Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
(WebKit::WebPageProxy::bindAccessibilityTree):
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEAccessibilityAtk.cpp: Added.
(WPE::initializeAtkUtil):
(WPE::accessibilityAtkInit):
* Source/WebKit/WPEPlatform/wpe/WPEAccessibilityAtk.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEApplicationAccessibleAtk.cpp: Added.
(wpeApplicationAccessibleAtkConstructed):
(wpeApplicationAccessibleAtkInitialize):
(wpeApplicationAccessibleAtkGetNChildren):
(wpeApplicationAccessibleAtkRefChild):
(wpeApplicationAccessibleAtkGetName):
(wpe_application_accessible_atk_class_init):
(wpeApplicationAccessibleAtkNew):
(wpeApplicationAccessibleAtkGetToplevelIndex):
* Source/WebKit/WPEPlatform/wpe/WPEApplicationAccessibleAtk.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEToplevel.cpp:
(wpeToplevelConstructed):
(wpeToplevelDispose):
(wpe_toplevel_class_init):
(wpeToplevelList):
(wpeToplevelGetView):
(wpeToplevelGetAccessibleAtk):
* Source/WebKit/WPEPlatform/wpe/WPEToplevelAccessibleAtk.cpp: Added.
(toplevelDestroyedCallback):
(wpeToplevelAccessibleAtkDispose):
(wpeToplevelAccessibleAtkInitialize):
(wpeToplevelAccessibleAtkRefStateSet):
(wpeToplevelAccessibleAtkGetIndexInParent):
(wpeToplevelAccessibleAtkGetNChildren):
(wpeToplevelAccessibleAtkRefChild):
(wpe_toplevel_accessible_atk_class_init):
(wpeToplevelAccessibleAtkGetSize):
(wpeToplevelAccessibleAtkComponentInterfaceInit):
(wpeToplevelAccessibleAtkWindowInterfaceInit):
(wpeToplevelAccessibleAtkNew):
* Source/WebKit/WPEPlatform/wpe/WPEToplevelAccessibleAtk.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEToplevelPrivate.h:
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpeViewGetAccessible):
(wpe_view_class_init):
(wpe_view_get_accessible):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/WPEViewAccessibleAtk.cpp: Added.
(viewDestroyedCallback):
(wpeViewAccessibleAtkDispose):
(wpeViewAccessibleAtkInitialize):
(wpeViewAccessibleAtkRefStateSet):
(wpeViewAccessibleAtkGetIndexInParent):
(wpe_view_accessible_atk_class_init):
(wpeViewAccessibleAtkBind):
(wpeViewAccessibleInterfaceInit):
(wpeViewAccessibleAtkNew):
* Source/WebKit/WPEPlatform/wpe/WPEViewAccessibleAtk.h: Added.
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:
* Tools/MiniBrowser/wpe/main.cpp:
(activate):

Canonical link: <a href="https://commits.webkit.org/292006@main">https://commits.webkit.org/292006@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9625d004ae0fec2bfc555489aa021b852b43a050

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94713 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14302 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4120 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99732 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45204 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14577 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22722 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/72246 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29549 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97715 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/85498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/52582 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/10550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44543 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/3308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101774 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21741 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/81242 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21988 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/81527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80623 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25182 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14967 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15194 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21718 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26831 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21379 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24848 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23118 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/wpe/WPEWebViewLegacy.cpp:
```diff
@@ -1,3 +1,4 @@
+
```
Odd empty line added here, please remove this change before landing.

carlosgcampos replied:
That's my keyboard doing weird things sometimes.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEApplicationAccessibleAtk.cpp:
```diff
@@ -0,0 +1,136 @@
+    auto* toplevel = accessible->priv->toplevels[index];
+    auto* toplevelAccessible = wpeToplevelGetOrCreateAccessibleAtk(toplevel);
+    if (!ATK_IS_OBJECT(toplevelAccessible))
+        return nullptr;
```
Can this really happen? I would thing that if the toplevel exists (which is the case, given at this point we know the index was correct), then it should always be possible to create an accessible object for it. Maybe `ASSERT(ATK_IS_OBJECT(toplevelAccessible))` instead?

carlosgcampos replied:
No it can't happen.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEApplicationAccessibleAtk.cpp:
```diff
@@ -0,0 +1,136 @@
+
+static const char* wpeApplicationAccessibleAtkGetName(AtkObject*)
+{
+    return g_get_prgname();
```
Would it make sense to use `g_application_get_application_id()` if there is a `GApplication` instance in use, and use `g_get_prgname()` as fallback?

carlosgcampos replied:
This is what all ATK implementations do, so better keep this for compatibility. I don't know what this is used for.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEApplicationAccessibleAtk.cpp:
```diff
@@ -0,0 +1,136 @@
+    atkObjectClass->get_n_children = wpeApplicationAccessibleAtkGetNChildren;
+    atkObjectClass->ref_child = wpeApplicationAccessibleAtkRefChild;
+    atkObjectClass->get_name = wpeApplicationAccessibleAtkGetName;
+    atkObjectClass->get_parent = nullptr;
```
Fun, I would imagine removing the method would fail, but it seems to work fine.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEApplicationAccessibleAtk.cpp:
```diff
@@ -0,0 +1,136 @@
+{
+    auto* toplevelAccessible = wpeToplevelGetOrCreateAccessibleAtk(toplevel);
+    if (!ATK_IS_OBJECT(toplevelAccessible))
+        return;
```
Same comment as above... Shouldn't creation of an accessible for toplevels always succed?

carlosgcampos replied:
Right

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEToplevel.cpp:
```diff
@@ -126,6 +175,29 @@ void wpeToplevelRemoveView(WPEToplevel* toplevel, WPEView* view)
+        return nullptr;
+
+    auto views = copyToVectorOf<GRefPtr<WPEView>>(toplevel->priv->views);
+    return views[index];
```
Creating a temporary vector to then extract a single items seems a bit wasteful. Why not returning the desired element directly?

Something like this:

```suggestion
    size_t current = 0;
    for (auto* view : toplevel->priv->views)
        if (current++ == index)
            return GRefPtr(view);
    ASSERT_NOT_REACHED();
```

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPEToplevel.cpp:
```diff
@@ -126,6 +175,29 @@ void wpeToplevelRemoveView(WPEToplevel* toplevel, WPEView* view)
 
+GRefPtr<WPEView> wpeToplevelGetView(WPEToplevel* toplevel, size_t index)
+{
+    if (index >= toplevel->priv->views.size())
+        return nullptr;
```
Why would the index ever be out of bounds? Shouldn't this be a `RELEASE_ASSERT()`?

carlosgcampos replied:
This is called from an ATK interface, we don't want to crash because some AT program called GetChild with a buggy index.

aperezdc reviewed (approved):
Looks good overall but I left a few comments with some doubts/ideas. PTAL before landing.

carlosgcampos committed:
[WPE] WPEPlatform: add default accessibility implementation using ATK
https://bugs.webkit.org/show_bug.cgi?id=289329

Reviewed by Adrian Perez de Castro.

When WebKit is built with ATK and the platform doesn't implement
WPEView::get_accessible, a default implementation is provided using ATK.

* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::accessible):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::~View):
(WKWPE::View::accessible const): Deleted.
* Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebViewLegacy.cpp:
(WKWPE::ViewLegacy::~ViewLegacy):
(WKWPE::ViewLegacy::accessible const):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewLegacy.h:
* Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
(WebKit::WebPageProxy::bindAccessibilityTree):
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEAccessibilityAtk.cpp: Added.
(WPE::initializeAtkUtil):
(WPE::accessibilityAtkInit):
* Source/WebKit/WPEPlatform/wpe/WPEAccessibilityAtk.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEApplicationAccessibleAtk.cpp: Added.
(wpeApplicationAccessibleAtkConstructed):
(wpeApplicationAccessibleAtkInitialize):
(wpeApplicationAccessibleAtkGetNChildren):
(wpeApplicationAccessibleAtkRefChild):
(wpeApplicationAccessibleAtkGetName):
(wpe_application_accessible_atk_class_init):
(wpeApplicationAccessibleAtkNew):
(wpeApplicationAccessibleAtkGetToplevelIndex):
* Source/WebKit/WPEPlatform/wpe/WPEApplicationAccessibleAtk.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEToplevel.cpp:
(wpeToplevelConstructed):
(wpeToplevelDispose):
(wpe_toplevel_class_init):
(wpeToplevelList):
(wpeToplevelGetView):
(wpeToplevelGetAccessibleAtk):
* Source/WebKit/WPEPlatform/wpe/WPEToplevelAccessibleAtk.cpp: Added.
(toplevelDestroyedCallback):
(wpeToplevelAccessibleAtkDispose):
(wpeToplevelAccessibleAtkInitialize):
(wpeToplevelAccessibleAtkRefStateSet):
(wpeToplevelAccessibleAtkGetIndexInParent):
(wpeToplevelAccessibleAtkGetNChildren):
(wpeToplevelAccessibleAtkRefChild):
(wpe_toplevel_accessible_atk_class_init):
(wpeToplevelAccessibleAtkGetSize):
(wpeToplevelAccessibleAtkComponentInterfaceInit):
(wpeToplevelAccessibleAtkWindowInterfaceInit):
(wpeToplevelAccessibleAtkNew):
* Source/WebKit/WPEPlatform/wpe/WPEToplevelAccessibleAtk.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEToplevelPrivate.h:
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpeViewGetAccessible):
(wpe_view_class_init):
(wpe_view_get_accessible):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/WPEViewAccessibleAtk.cpp: Added.
(viewDestroyedCallback):
(wpeViewAccessibleAtkDispose):
(wpeViewAccessibleAtkInitialize):
(wpeViewAccessibleAtkRefStateSet):
(wpeViewAccessibleAtkGetIndexInParent):
(wpe_view_accessible_atk_class_init):
(wpeViewAccessibleAtkBind):
(wpeViewAccessibleInterfaceInit):
(wpeViewAccessibleAtkNew):
* Source/WebKit/WPEPlatform/wpe/WPEViewAccessibleAtk.h: Added.
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:
* Tools/MiniBrowser/wpe/main.cpp:
(activate):

Canonical link: https://commits.webkit.org/292006@main
